File: netCDF/examples/C/example_par_read.c
#include "mpi.h"
#include <netcdf.h>
#include <netcdf_par.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NDIMS 2

#define ERR(e) {printf("Error: %s\n", nc_strerror(e)); return 2;}

// Function decomposes global domain of size ni_glo*nj_glo in horizontal direction into local domains
void decompose_domain(int mpisize, int mpirank, int ni_glo, int nj_glo, int* ni, int* ibegin, int* nj, int* jbegin)
{
    *nj=nj_glo;
    *ibegin=0;
    *jbegin=0;
    for (int n=0; n<mpisize; ++n)
    {     
      *ni=ni_glo/mpisize;
      if (n<(ni_glo%mpisize)) ++(*ni);
      if (n==mpirank) break;
      (*ibegin)+=(*ni);
    }
}

// Funciton checks the correctness of the read data
void check_read_data(int* data, int ni, int ibegin, int nj, int jbegin)
{
  int ok = 0;
  for(int i=0; i < ni; i++)
    for(int j=0; j < nj; j++)
      if (data[j+i*nj] != (j+jbegin) + (i+ibegin)*nj)
        ++ok;
  if (ok == 0)
    printf("Parallel read is done successfully.\n");
  else
    printf("Error: something went wrong; data read from the file do not correspond to the expected.\n");
}

int main (void)
{
  size_t starts[NDIMS], counts[NDIMS];
  int mpirank,mpisize;
  int ncid,nidimid,njdimid,varid;
  int err;
  size_t ni_glo, nj_glo;
  int ni, nj, ibegin, jbegin;
  char filename[128] = "par_write_.nc";

  //--------------------------------------
  // Initialization MPI
  //--------------------------------------

  MPI_Init(NULL, NULL);
  MPI_Comm_rank(MPI_COMM_WORLD,&mpirank);
  MPI_Comm_size(MPI_COMM_WORLD,&mpisize);

  //--------------------------------------
  //  NetCDF metadata definitions
  //--------------------------------------

  // Open netcdf file
  if (err=nc_open_par(filename, NC_NOWRITE, MPI_COMM_WORLD, MPI_INFO_NULL, &ncid)) ERR(err);

  // Get IDs of two spatial dimensions given their names
  if (err=nc_inq_dimid(ncid, "ni",  &nidimid)) ERR(err);
  if (err=nc_inq_dimid(ncid, "nj",  &njdimid)) ERR(err);

  // Get two spatial dimensions
  if (err=nc_inq_dimlen(ncid, nidimid, &ni_glo)) ERR(err);
  if (err=nc_inq_dimlen(ncid, njdimid, &nj_glo)) ERR(err);

  // Get ID of the spacial variable given its name
  if (err=nc_inq_varid(ncid, "toto",  &varid)) ERR(err);

  // Decompose global domain
  // Returned values: ni,nj,ibegin,jbegin
  decompose_domain(mpisize,mpirank,ni_glo, nj_glo, &ni,&ibegin,&nj,&jbegin);
  int *data = (int *)malloc(sizeof(int)*ni*nj);

  //--------------------------------------
  // Read data and close the file  
  //--------------------------------------

  // Read data
  starts[0] = ibegin; starts[1] = jbegin;
  counts[0] = ni;     counts[1] = nj;
  if (err=nc_get_vara(ncid, varid, starts, counts, data)) ERR(err);
  
  // Close the file
  if (err=nc_close(ncid)) ERR(err);

  // Verify the read data
  check_read_data(data, ni, ibegin, nj, jbegin);  

  free(data);

  MPI_Finalize();

  return 0;
}     

File: netCDF/examples/F90/example_par_read.f90
program example_par_read
  use netcdf
  implicit none
  include 'mpif.h'

  integer :: mpirank
  integer :: mpisize
  integer :: ierr

  character (len = *), parameter :: FILE_NAME = "par_write.nc"
  integer,parameter :: NDIMS=2
  integer :: ni_glo, nj_glo
  integer,allocatable :: field(:,:)
  integer :: ni,ibegin,nj,jbegin,i,j
  integer :: ncid,nidimid,njdimid,varid

  ! MPI initialization
  call MPI_Init(ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD,mpirank,ierr)
  call MPI_Comm_size(MPI_COMM_WORLD,mpisize,ierr)

  ! Open the file
  call check( nf90_open(FILE_NAME, NF90_NOWRITE, ncid,&
                    comm=MPI_COMM_WORLD, info=MPI_INFO_NULL))

  ! Get IDs of the two spatial dimensions given their names
  call check( nf90_inq_dimid(ncid, "ni", nidimid) )
  call check( nf90_inq_dimid(ncid, "nj", njdimid) )

  ! Get two spatial dimensions
  call check( nf90_inquire_dimension(ncid, nidimid, len = ni_glo) )
  call check( nf90_inquire_dimension(ncid, njdimid, len = nj_glo) )

  ! Get ID of the spacial variable given its name
  call check( nf90_inq_varid(ncid, "toto", varid) )

  ! Decompose global domain.
  ! Returned values: ni,nj,ibegin,jbegin.
  call decompose_domain(mpisize,mpirank,ni_glo,nj_glo,ni,ibegin,nj,jbegin)

  ! Allocate local data array
  allocate(field(ni,nj))

  ! Read data
  call check( nf90_get_var(ncid, varid, field, start=(/ibegin,jbegin/), count=(/ni,nj/)) )

  ! Close the file
  call check( nf90_close(ncid) )

  ! Verify the read data
  do j = 1, nj
    do i = 1, ni
      if (field(i, j) /= (j-1+jbegin-1)*ni + (i-1+ibegin-1)) then
        print *, "Error: something went wrong; data read from the file do not correspond to the expected."
        stop 2 
      endif
    end do
  end do
  
  print *, "Parallel read is done successfully."

  deallocate(field)

  call MPI_FINALIZE(ierr)

  !--------------------------------------
  ! Functions and subroutines
  !--------------------------------------

  contains 

  subroutine check(status)
    integer, intent ( in) :: status

    if(status /= nf90_noerr) then
      print *, trim(nf90_strerror(status))
      stop 2
    end if
  end subroutine check

  !--------------------------------------
  ! This subroutine decomposes global domain
  !--------------------------------------
  subroutine decompose_domain(mpisize,rank,ni_glo,nj_glo,ni,ibegin,nj,jbegin)
    integer,intent(in) :: mpisize,rank,ni_glo,nj_glo
    integer,intent(out) :: ni,ibegin,nj,jbegin
    integer :: n

    ni=ni_glo ; ibegin=1
    jbegin=0
    do n=0,mpisize-1
      nj=nj_glo/mpisize
      if (n<MOD(nj_glo,mpisize)) nj=nj+1
      if (n==rank) exit
      jbegin=jbegin+nj
    enddo
    jbegin=jbegin+1
  end subroutine decompose_domain

end program example_par_read

